spec/lib/api/contracts/model_contract_spec.rb

```diff
@@ -0,0 +1,89 @@
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+
+require 'spec_helper'
+
+describe ::API::Contracts::ModelContract do
+  let(:work_package) { FactoryGirl.create(:work_package) }
+  let(:model_child_contract) { ModelChild.new(work_package) }
+  let(:model_grand_child_contract) { ModelGrandChild.new(work_package) }
+
+  before do
+    model_child_contract.child_value = 0
+    model_grand_child_contract.child_value = 0
+  end
+
+  describe 'child' do
+    class ModelChild < ::API::Contracts::ModelContract
+      attr_accessor :child_value
+
+      attribute :child_attribute
+      attribute :overwritten_attribute do
+        @child_value = 1
+      end
+    end
+
+    it 'should collect its own writable attributes' do
+      expect(model_child_contract.writable_attributes).to include('child_attribute',
+                                                                  'overwritten_attribute')
+    end
+
+    it 'should collect its own attribute validations' do
+      model_child_contract.validate
+      expect(model_child_contract.child_value).to eq(1)
+    end
+  end
+
+  describe 'grand_child' do
+    class ModelGrandChild < ModelChild
+      attr_accessor :grand_child_value
+
+      attribute :grand_child_attribute
+      attribute :overwritten_attribute do
+        @grand_child_value = 2
+      end
+    end
+
+    it 'should consider its ancestor writable attributes' do
+      expect(model_grand_child_contract.writable_attributes).to include('child_attribute',
+                                                                        'overwritten_attribute',
+                                                                        'grand_child_attribute')
+      expect(model_grand_child_contract.writable_attributes.count).to eq(3)
+    end
+
+    it 'should obey common inheritance rules' do
+      expect(model_grand_child_contract.writable_attributes.count).to eq(3)
```

Comment:
Looks like more remnants. The test above should not include his one. The common inheritance rule this test is talking about revolves around the ``overwritten_attribute`` but since this test is just checking inheritance in general it can probably be removed anyway.